runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2984,48 +2986,46 @@ J9::ValuePropagation::isArrayCompTypePrimitiveValueType(TR::VPConstraint *arrayC
    if (!(arrayConstraint && arrayConstraint->getClass()
               && arrayConstraint->getClassType()->isArray() == TR_yes))
       {
+      if (trace())
+         traceMsg(comp(), "%s: return TR_maybe. arrayConstraint %p\n", __FUNCTION__, arrayConstraint);
       return TR_maybe;
       }
 
-   TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayConstraint->getClass());
-
-   // Cases to consider:
-   //
-   //   - Is no information available about the component type of the array?
-   //     If not, assume it might be a primitive value type.
-   //   - Is the component type definitely a identity type?
-   //   - Is the component type definitely a primitive value type?
-   //   - Is the component type definitely a value type, but not primitive?
-   //   - Is the component type either an abstract class or an interface
-   //     (i.e., not a concrete class)?  If so, it might be a value type.
-   //   - Is the array an array of java/lang/Object?  See below.
-   //   - Otherwise, it must be a concrete class known not to be a value
-   //     type
-   //
-   if (!arrayComponentClass)
-      {
-      return TR_maybe;
-      }
-
-   // No need to check array class type because array classes should be marked as having identity.
-   if (TR::Compiler->cls.classHasIdentity(arrayComponentClass))
-      {
-      return TR_no;
-      }
+   TR_OpaqueClassBlock *arrayClass = arrayConstraint->getClass();
 
-   if (TR::Compiler->cls.isPrimitiveValueTypeClass(arrayComponentClass))
+   if (TR::Compiler->cls.isArrayNullRestricted(comp(), arrayClass))
       {
+      if (trace())
+         traceMsg(comp(), "%s: return TR_yes. arrayClass isArrayNullRestricted\n", __FUNCTION__);
       return TR_yes;
       }
 
-   if (TR::Compiler->cls.isValueTypeClass(arrayComponentClass))
+   TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayConstraint->getClass());
+
+   if (!arrayComponentClass)
       {
-      return TR_no;
+      if (trace())
+         traceMsg(comp(), "%s: return TR_maybe. arrayComponentClass NULL\n", __FUNCTION__);
+      return TR_maybe;
       }
 
    if (!TR::Compiler->cls.isConcreteClass(comp(), arrayComponentClass))
       {
-      return TR_maybe;
+      // Interface shouldn't have identity flag set and it can be subclassed by both
+      // value class and identity class.
+      // If abstract class has identity flag set, it cannot be subclassed by value class.
```

Comment:
A couple of nits:  "subclassed by" should be "implemented by" for an interface and "extended by" for an abstract class